src/java/picard/analysis/CollectWgsMetricsFromQuerySorted.java

```diff
@@ -0,0 +1,300 @@
+
+            final boolean isProperPair = (pairToAnalyze.read2 != null);
+
+            // how many bases do we have?
+            final int read1bases = pairToAnalyze.read1.getReadLength();
+            final int read2bases = isProperPair ? pairToAnalyze.read2.getReadLength() : 0;
+            final int totalReadBases = read1bases + read2bases;
+
+            // now compute metrics...
+            metrics.TOTAL_BASES += totalReadBases;
+
+            if (!isProperPair || pairToAnalyze.read1.getMateUnmappedFlag() || pairToAnalyze.read2.getMateUnmappedFlag()) {
+                basesExcludedByPairing += totalReadBases;
+            } else if (pairToAnalyze.read1.getDuplicateReadFlag()) {
+                basesExcludedByDupes += totalReadBases;
+            } else {
+
+                // determine the bad bases from the reads
+                final BaseExclusionHelper read1exclusions = determineBaseExclusions(pairToAnalyze.read1);
+                final BaseExclusionHelper read2exclusions = determineBaseExclusions(pairToAnalyze.read2);
+                basesExcludedByMapq += read1exclusions.basesExcludedByMapq + read2exclusions.basesExcludedByMapq;
+                basesExcludedByBaseq += read1exclusions.lowBQcount + read2exclusions.lowBQcount;
+
+                // keep track of the total usable bases
+                int usableBaseCount = totalReadBases;
+                usableBaseCount -= (read1exclusions.basesExcludedByMapq + read1exclusions.lowBQcount);
+                usableBaseCount -= (read2exclusions.basesExcludedByMapq + read2exclusions.lowBQcount);
+
+                // subtract out bad bases from overlaps between the reads, but only if both reads pass mapping quality thresholds
+                if (read1exclusions.basesExcludedByMapq == 0 && read2exclusions.basesExcludedByMapq == 0) {
+                    final int overlapCount = getOverlappingBaseCount(read1exclusions, read2exclusions);
+                    basesExcludedByOverlap += overlapCount;
+                    usableBaseCount -= overlapCount;
+                }
+
+                metrics.TOTAL_USABLE_BASES += usableBaseCount;
+            }
+
+            // record progress
+            progress.record(pairToAnalyze.read1);
+
+            // iterate
+            pairToAnalyze = getNextReadPair(iterator);
+        }
+
+        // finalize metrics
+        metrics.GENOME_TERRITORY = reader.getFileHeader().getSequenceDictionary().getReferenceLength();
+        metrics.MEAN_COVERAGE = metrics.TOTAL_USABLE_BASES / (double)metrics.GENOME_TERRITORY;
+        metrics.PCT_EXC_DUPE = basesExcludedByDupes / (double)metrics.TOTAL_BASES;
+        metrics.PCT_EXC_MAPQ = basesExcludedByMapq / (double)metrics.TOTAL_BASES;
+        metrics.PCT_EXC_UNPAIRED = basesExcludedByPairing / (double)metrics.TOTAL_BASES;
+        metrics.PCT_EXC_BASEQ = basesExcludedByBaseq / (double)metrics.TOTAL_BASES;
+        metrics.PCT_EXC_OVERLAP = basesExcludedByOverlap / (double)metrics.TOTAL_BASES;
```

Comment:
You should probably give an invalid value (-1? NaN? ) to the fields you are not filling since otherwise they will have the valid value of zero, and it could be confusing.